core/src/upgrade/secure.rs

```diff
@@ -0,0 +1,88 @@
+//! After a successful protocol negotiation as part of the upgrade process, the `SecurityUpgrade::upgrade_security`
+//! method is called and a [`Future`] that performs a handshake is returned.
+
+use crate::upgrade::{SecurityUpgrade, UpgradeError};
+use crate::{connection::ConnectedPoint, Negotiated};
+use futures::prelude::*;
+
+use libp2p_identity::PeerId;
+use multiaddr::Protocol;
+use multistream_select::Version;
+
+/// Applies a security upgrade to the inbound and outbound direction of a connection or substream.
+pub(crate) async fn secure<C, U>(
+    conn: C,
+    up: U,
+    cp: ConnectedPoint,
+    v: Version,
+) -> Result<U::Output, UpgradeError<U::Error>>
+where
+    C: AsyncRead + AsyncWrite + Unpin,
+    U: SecurityUpgrade<Negotiated<C>>,
+{
+    match cp {
+        ConnectedPoint::Dialer { role_override, .. } if role_override.is_dialer() => {
+            let peer_id = cp
+                .get_remote_address()
+                .iter()
+                .find_map(|protocol| match protocol {
+                    Protocol::P2p(peer_id) => Some(peer_id),
+                    _ => None,
+                })
+                .expect("It should have /p2p as part of the address");
```

Comment:
Can't panic here, if it is not there we should set it to `None`.

Reply:
Yes, I forgot about the panic.